onnxscript/rewriter/custom_ops/llm_rule_sets_cuda.py

```diff
@@ -0,0 +1,495 @@
+    def pattern(cls, op, shape, indices, updates, tensor, masked, zero, reduction):
+        cst = op.ConstantOfShape(shape, value=tensor)
+        masked_indices = op.Equal(indices, masked)
+        masked_updates = op.Where(masked_indices, zero, updates)
+        return op.ScatterND(cst, indices, masked_updates, reduction=reduction)
+
+    @classmethod
+    def check(cls, context, shape, indices, updates, tensor, masked, zero, reduction) -> bool:
+        if reduction.value != "add":
+            return False
+        if tensor.value.numpy().reshape((1,)).tolist() != [0]:
+            return False
+        if zero.const_value is None or zero.const_value.numpy().reshape((1,)).tolist() != [0]:
+            return False
+        if masked.const_value is None or masked.const_value.numpy().size != 1:
+            return False
+        return True
+
+    @classmethod
+    def rewrite(cls, op, shape, indices, updates, tensor, masked, zero, reduction):
+        return op.MaskedScatterNDOfShape(
+            shape,
+            indices,
+            updates,
+            maskedValue=int(masked.const_value.numpy().reshape((1,))[0]),
+            reduction=reduction.value,
+            domain="ai.onnx.contrib",
+        )
+
+
+class MulSigmoid(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, x):
+        return op.Mul(x, op.Sigmoid(x))
+
+    @classmethod
+    def check(cls, context, x) -> bool:
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x):
+        return op.MulSigmoid(x, domain="ai.onnx.contrib")
+
+
+class NegXPlus1(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, cst, x):
+        return op.Sub(cst, x)
+
+    @classmethod
+    def check(cls, context, cst, x) -> bool:
+        if cst.const_value is None:
+            return False
+        if cst.shape != (1,):
+            return False
+        value = float(cst.const_value.numpy().reshape((1,))[0])
+        return value == 1
+
+    @classmethod
+    def rewrite(cls, op, cst, x):
+        return op.NegXplus1(x, domain="ai.onnx.contrib")
+
+
+class ReplaceZero1(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, x, y):
+        return op.Where(op.Cast(x, to=onnx.TensorProto.BOOL), x, y)
+
+    @classmethod
+    def check(cls, context, x, y) -> bool:
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x, y):
+        return op.ReplaceZero(x, y, equal=1, domain="ai.onnx.contrib")
+
+
+class ReplaceZero2(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, x, y):
+        return op.Where(op.Cast(x, to=onnx.TensorProto.BOOL), y, x)
+
+    @classmethod
+    def check(cls, context, x, y) -> bool:
+        return True
+
+    @classmethod
+    def rewrite(cls, op, x, y):
+        return op.ReplaceZero(x, y, equal=0, domain="ai.onnx.contrib")
+
+
+class Rotary1(orp.RewriteRuleAsClass):
+    @classmethod
+    def pattern(cls, op, x):
+        x1, x2 = op.Split(x, axis=-1, outputs=2, num_outputs=2)
```

Comment:
This reminds me: I think the generic pattern matcher doesn't check for attribute-values like "axis=-1". It is easy to add (the underlying pattern object has a match method to do it), but I think this is still a to-do. Trying to understand whether you want the check here? If so, it can't be working right now?

Reply:
You are right, it does not. I forgot to add that verification.

Reply:
done